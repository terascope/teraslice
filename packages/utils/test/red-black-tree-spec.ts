import 'jest-extended';
import {
    RedBlackTree
} from '../src/red-black-tree';

describe('RedBlackTree', () => {
    describe('when adding one item', () => {
        let tree: RedBlackTree<string>;
        beforeAll(() => {
            tree = new RedBlackTree();
            tree.insert(2, 'hi');
        });

        it('should have written the root node', () => {
            expect(tree.root).toMatchSnapshot();
        });
    });

    describe('when adding three items', () => {
        let tree: RedBlackTree<string>;
        beforeAll(() => {
            tree = new RedBlackTree();
            tree.insert(2, 'hi');
            tree.insert(3, 'hello');
            tree.insert(10, 'howdy');
        });

        it('should have written a tree of nodes', () => {
            expect(tree.root).toMatchSnapshot();
        });
    });

    describe('when adding 10 items', () => {
        let tree: RedBlackTree<string>;
        beforeAll(() => {
            tree = new RedBlackTree();
            tree.insert(5, 'hello');
            tree.insert(18, 'howdy');
            tree.insert(2, 'hi');
            tree.insert(10, 'howdy there');
            tree.insert(200, 'hello friend');
            tree.insert(60, 'aloha');
            tree.insert(30, 'hello person');
            tree.insert(90, 'sup');
            tree.insert(3, 'hi');
            tree.insert(33, 'oh hey');
        });

        it('should have written a tree of nodes', () => {
            expect(tree.root).toMatchSnapshot();
        });
    });

    describe('when adding 10 items and some values are null', () => {
        let tree: RedBlackTree<string>;
        beforeAll(() => {
            tree = new RedBlackTree();
            tree.insert(5, 'hello');
            tree.insert(18, 'howdy');
            tree.insert(2, 'hi');
            tree.insert(10, null);
            tree.insert(200, null);
            tree.insert(60, 'aloha');
            tree.insert(30, null);
            tree.insert(90, 'sup');
            tree.insert(3, 'hi');
            tree.insert(33, null);
        });

        it('should have written a tree of nodes', () => {
            expect(tree.root).toMatchSnapshot();
        });
    });
});
