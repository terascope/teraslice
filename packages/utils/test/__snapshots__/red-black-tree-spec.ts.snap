// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RedBlackTree when adding 10 items and some values are null should have written a tree of nodes 1`] = `
Object {
  "color": "RED",
  "key": 5,
  "left": Object {
    "color": "RED",
    "key": 2,
    "left": undefined,
    "right": Object {
      "color": "BLACK",
      "key": 3,
      "left": undefined,
      "right": undefined,
      "value": "hi",
    },
    "value": "hi",
  },
  "right": Object {
    "color": "BLACK",
    "key": 18,
    "left": Object {
      "color": "BLACK",
      "key": 10,
      "left": undefined,
      "right": undefined,
      "value": null,
    },
    "right": Object {
      "color": "RED",
      "key": 200,
      "left": Object {
        "color": "RED",
        "key": 60,
        "left": Object {
          "color": "BLACK",
          "key": 30,
          "left": undefined,
          "right": Object {
            "color": "BLACK",
            "key": 33,
            "left": undefined,
            "right": undefined,
            "value": null,
          },
          "value": null,
        },
        "right": Object {
          "color": "BLACK",
          "key": 90,
          "left": undefined,
          "right": undefined,
          "value": "sup",
        },
        "value": "aloha",
      },
      "right": undefined,
      "value": null,
    },
    "value": "howdy",
  },
  "value": "hello",
}
`;

exports[`RedBlackTree when adding 10 items should have written a tree of nodes 1`] = `
Object {
  "color": "RED",
  "key": 5,
  "left": Object {
    "color": "RED",
    "key": 2,
    "left": undefined,
    "right": Object {
      "color": "BLACK",
      "key": 3,
      "left": undefined,
      "right": undefined,
      "value": "hi",
    },
    "value": "hi",
  },
  "right": Object {
    "color": "RED",
    "key": 18,
    "left": Object {
      "color": "BLACK",
      "key": 10,
      "left": undefined,
      "right": undefined,
      "value": "howdy there",
    },
    "right": Object {
      "color": "RED",
      "key": 200,
      "left": Object {
        "color": "RED",
        "key": 60,
        "left": Object {
          "color": "RED",
          "key": 30,
          "left": undefined,
          "right": Object {
            "color": "BLACK",
            "key": 33,
            "left": undefined,
            "right": undefined,
            "value": "oh hey",
          },
          "value": "hello person",
        },
        "right": Object {
          "color": "BLACK",
          "key": 90,
          "left": undefined,
          "right": undefined,
          "value": "sup",
        },
        "value": "aloha",
      },
      "right": undefined,
      "value": "hello friend",
    },
    "value": "howdy",
  },
  "value": "hello",
}
`;

exports[`RedBlackTree when adding one item should have written the root node 1`] = `
Object {
  "color": "BLACK",
  "key": 2,
  "left": undefined,
  "right": undefined,
  "value": "hi",
}
`;

exports[`RedBlackTree when adding three items should have written a tree of nodes 1`] = `
Object {
  "color": "RED",
  "key": 2,
  "left": undefined,
  "right": Object {
    "color": "RED",
    "key": 3,
    "left": undefined,
    "right": Object {
      "color": "BLACK",
      "key": 10,
      "left": undefined,
      "right": undefined,
      "value": "howdy",
    },
    "value": "hello",
  },
  "value": "hi",
}
`;
